invoice/invoice.go

```diff
@@ -0,0 +1,966 @@
+	default:
+		return nil, fmt.Errorf("unknown network: %v", hrp[2:4])
+	}
+	decodedInvoice.Net = net
+
+	// Optionally, if there's anything left of the HRP, it encodes the
+	// payment amount.
+	if len(hrp) > 4 {
+		amount, err := decodeAmount(hrp[4:])
+		if err != nil {
+			return nil, err
+		}
+		decodedInvoice.MilliSat = &amount
+	}
+
+	// Everything except the last 520 bits of the data encodes the invoice's
+	// timestamp and tagged fields.
+	invoiceData := data[:len(data)-104]
+
+	// Parse the timestamp and tagged fields, and fill the Invoice struct.
+	if err := parseData(&decodedInvoice, invoiceData, net); err != nil {
+		return nil, err
+	}
+
+	// The last 520 bits (104 groups) make up the signature.
+	sigBase32 := data[len(data)-104:]
+	sigBase256, err := bech32.ConvertBits(sigBase32, 5, 8, true)
+	if err != nil {
+		return nil, err
+	}
+	var sigBytes [64]byte
+	copy(sigBytes[:], sigBase256[:64])
+	recoveryID := sigBase256[64]
+
+	// The signature is over the hrp + the data the invoice, encoded in
+	// base 256.
+	taggedDataBytes, err := bech32.ConvertBits(invoiceData, 5, 8, true)
+	if err != nil {
+		return nil, err
+	}
+
+	toSign := append([]byte(hrp), taggedDataBytes...)
+
+	// We expect the signature to be over the single SHA-256 hash of that
+	// data.
+	hash := chainhash.HashB(toSign)
+
+	// If pubkey was provided as a tagged field, use that to verify the
+	// signature, if not do public key recovery.
+	if decodedInvoice.PubKey != nil {
+		var signature *btcec.Signature
+		lnwire.DeserializeSigFromWire(&signature, sigBytes)
+		if !signature.Verify(hash, decodedInvoice.PubKey) {
+			return nil, fmt.Errorf("invalid invoice signature")
+		}
+	} else {
+		headerByte := recoveryID + 27 + 4
+		compactSign := append([]byte{headerByte}, sigBytes[:]...)
+		pubkey, _, err := btcec.RecoverCompact(btcec.S256(),
+			compactSign, hash)
+		if err != nil {
+			return nil, err
+		}
+		decodedInvoice.RecoveredPubKey = pubkey
+	}
+
+	// Now that we have created the invoice, make sure it has the required
+	// fields set.
+	if err := validateInvoice(&decodedInvoice); err != nil {
+		return nil, err
+	}
+
+	return &decodedInvoice, nil
+}
+
+// Encode takes the given MessageSigner and returns a string encoding this
+// invoice signed by the node key of the signer.
+func (invoice *Invoice) Encode(signer MessageSigner) (string, error) {
+	// First check that this invoice is valid before starting the encoding.
+	if err := validateInvoice(invoice); err != nil {
+		return "", err
+	}
+
+	// The buffer will encoded the invoice data using 5-bit groups (base32).
+	var bufferBase32 bytes.Buffer
+
+	// The timestamp will be encoded using 35 bits, in base32.
+	timestampBase32 := uint64ToBase32(invoice.Timestamp)
+
+	// The timestamp must be exactly 35 bits, which means 7 groups. If it
+	// can fit into fewer groups we add leading zero groups, if it is too
+	// big we fail early, as there is not possible to encode it.
+	if len(timestampBase32) > 7 {
```

Comment:
Similar comment here about extracting the `7` value to a constant above with a `godoc` comment explaining the required grouping.